File: 프로그래머스/3/68646. 풍선 터트리기/풍선 터트리기.java
class Solution {
    public int solution(int[] a) {
        int[] leftMin = new int[a.length]; //현재 위치에서 왼쪽 값들 중 가장 작은 값
        int[] rightMin = new int[a.length]; //현재 위치에서 오른쪽 값들 중 가장 작은 값
        int L = a[0];
        int R = a[a.length-1];
        
        //최솟값 저장
        for(int i=1; i<a.length-1; i++){
            if(L > a[i]) L = a[i];
            leftMin[i] = L;
        }
        
        //최댓값 저장
        for(int i=a.length-2; i>0; i--){
            if(R > a[i]) R = a[i];
            rightMin[i] = R;
        }
        
        if(a.length == 1) return 1;
        int answer = 2; //배열 a의 양 끝 값은 항상 true임.
        for(int i=1; i<a.length-1; i++){
            if(a[i] > leftMin[i] && a[i] > rightMin[i]) continue;
            answer ++;
        }
        return answer;
    }
}

File: 프로그래머스/3/68646. 풍선 터트리기/풍선 터트리기.kt
class Solution {
    fun solution(a: IntArray): Int {
        //양쪽 끝 값은 항상 최후까지 남을 수 있음.
        var answer: Int = 2
        
        var leftMin = IntArray(a.size)
        var rightMin = IntArray(a.size)
        
        var left = a[0]
        var right = a[a.size-1]
        for(i in 1 until a.size-1){
            left = minOf(left, a[i])
            leftMin[i] = left
        }
        
        for(i in a.size-1 downTo 0){
            right = minOf(right, a[i])
            rightMin[i] = right
        }
        
        for(i in 1 until a.size-1){
            //i 기준 양 쪽 값이 둘 다 크면 ans 증가
            if(a[i] < leftMin[i] && a[i] < rightMin[i]) {
                answer++
            //i 기준 양 쪽 값이 둘 다 작으면
            }else if(a[i] > leftMin[i] && a[i] > rightMin[i]){
                continue
            //i 기준 양 쪽 값들 중 하나만 작으면
            }else{
                answer++
            }
        }
        
        return answer
    }
}